File: leetcode/algorithms/111_minimum_depth_of_binary_tree/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * Recursion: DFS
 *
 * Complexities:
 *   N - Number of nodes in `root`
 *   H - Width of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 *
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
  if (root === null) {
    return 0;
  }

  const leftDepth = minDepth(root.left);
  const rightDepth = minDepth(root.right);

  if (leftDepth === 0 || rightDepth === 0) {
    return Math.max(leftDepth, rightDepth) + 1;
  }

  return Math.min(leftDepth, rightDepth) + 1;
};


// Solution
/**
 * Solution 1
 *
 * Iteration: BFS (Queue)
 *
 * Complexities:
 *   N - Number of nodes in `root`
 *   W - Width of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(W)
 *
 * @param {TreeNode} root
 * @return {number}
 */
var solution1 = function (root) {
  if (!root) {
    return 0;
  }

  const queue = [root];
  let depth = 1;

  while (queue.length > 0) {
    const levelSize = queue.length;

    for (let i = 0; i < levelSize; i++) {
      const node = queue.shift();

      if (node.left === null && node.right === null) {
        return depth;
      }

      if (node.left) {
        queue.push(node.left);
      }
      if (node.right) {
        queue.push(node.right);
      }
    }

    depth++;
  }

  return depth;
};

/**
 * Solution 2
 *
 * Recursion: DFS
 *
 * Complexities:
 *   N - Number of nodes in `root`
 *   H - Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 *
 * @param {TreeNode} root
 * @return {number}
 */
var solution2 = function (root) {
  if (!root) {
    return 0;
  }

  if (root.left === null) {
    return solution2(root.right) + 1;
  }

  if (root.right === null) {
    return solution2(root.left) + 1;
  }

  return Math.min(solution2(root.left), solution2(root.right)) + 1;
};

File: leetcode/algorithms/111_minimum_depth_of_binary_tree/main.py
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class MinimumDepthOfBinaryTree:
    """
    # Recursion: DFS
    #
    # Complexities:
    #   N - Number of nodes in `root`
    #   H - Height of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(H)
    """
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        left_depth: int = self.minDepth(root.left)
        right_depth: int = self.minDepth(root.right)

        if left_depth == 0 or right_depth == 0:
            return max(left_depth, right_depth) + 1

        return min(left_depth, right_depth) + 1


    # Solution
    """
    # Solution 1
    #
    # Iteration: BFS (Queue)
    #
    # Complexities:
    #   N - Number of nodes in `root`
    #   W - Width of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(W)
    """
    def solution1(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        queue = deque([(root, 1)])
        
        while queue:
            node, depth = queue.popleft()
            
            if not node.left and not node.right:
                return depth
            
            if node.left:
                queue.append((node.left, depth + 1))
            if node.right:
                queue.append((node.right, depth + 1))
                
        return 0

    """
    # Solution 2
    #
    # Recursion: DFS
    #
    # Complexities:
    #   N - Number of nodes in `root`
    #   H - Height of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(H)
    """
    def solution2(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        if not root.left and not root.right:
            return 1
        
        if not root.left:
            return self.solution2(root.right) + 1
        if not root.right:
            return self.solution2(root.left) + 1
        
        return min(self.solution2(root.left), self.solution2(root.right)) + 1
